File: client/packet.ts
import * as ProtoBuf from "protobufjs";
import * as ByteBuffer from "bytebuffer";
const root = ProtoBuf.loadSync("./test.proto");

class packet {
    protected message: any;
    protected pack: any;

    public decode(data: any): [string, any] {
        // 解包
        const bufferArray: Uint8Array = new Uint8Array(data);
        const buffer: ByteBuffer = new ByteBuffer().append(bufferArray);
        const headLen = buffer.readShort(0);
        const cmd = buffer.readString(headLen, 2, 0);
        let resp: any = root.lookupType(`test.${cmd.string}`);
        return (
            cmd.string,
            resp.decode(
                buffer.buffer.slice(2 + cmd.string.length, bufferArray.length)
            )
        );
    }

    public encode(cmd: string, data: any): object {
        let req = root.lookupType(`test.${cmd}`);
        // 设置协议内容
        let pack = req.create(data);
        let buff = req.encode(pack).finish();
        // 打包
        let buffer = new ByteBuffer(cmd.length + 2 + buff.length);
        buffer.writeShort(cmd.length);
        buffer.writeString(cmd);
        buffer.append(buff);
        return buffer.buffer;
    }
}

const Packet = new packet();
export default Packet;

File: server/utils/packet/packet.go
package packet

import (
	"bytes"
	"encoding/binary"
)

func Encode(cmd string, p []byte) []byte {
	buffer := &bytes.Buffer{}
	// 2字节，协议名称长度
	buffer.Write(shortBytes(uint16(len(cmd))))
	// 协议名称
	buffer.WriteString(cmd)
	// 协议内容
	buffer.Write(p)
	return buffer.Bytes()
}

func Decode(p []byte) (string, []byte) {
	buffer := bytes.NewBuffer(p)
	// 2字节，协议名称长度
	headLen := make([]byte, 2)
	buffer.Read(headLen)
	// 协议名称
	cmdLen := readShort(headLen)
	cmd := make([]byte, cmdLen)
	buffer.Read(cmd)
	// 协议内容
	var bytes []byte
	if l := buffer.Len(); l > 0 {
		bytes = make([]byte, l)
		buffer.Read(bytes)
	}
	return string(cmd), bytes
}

func readShort(b []byte) uint16 {
	return binary.BigEndian.Uint16(b)
}

func shortBytes(i uint16) []byte {
	bytes := make([]byte, 2)
	binary.BigEndian.PutUint16(bytes, i)
	return bytes
}
